File: day8_part2_2018.dart

import 'dart:io';

class Node {
  List<Node> children = [];
  List<int> metadata = [];
  
  int sumMetadata() {
    int sum = metadata.reduce((value, element) => value + element);
    for (var child in children) {
      sum += child.sumMetadata();
    }
    return sum;
  }
  
  int calculateValue() {
    if (children.isEmpty) {
      return metadata.reduce((value, element) => value + element);
    } else {
      int value = 0;
      for (var index in metadata) {
        if (index > 0 && index <= children.length) {
          value += children[index - 1].calculateValue();
        }
      }
      return value;
    }
  }
}

Node parseNode(List<int> data, int index) {
  Node node = Node();
  int numChildren = data[index];
  int numMetadata = data[index + 1];
  int i = index + 2;
  
  for (int j = 0; j < numChildren; j++) {
    Node child = parseNode(data, i);
    node.children.add(child);
    i += childSize(child);
  }
  
  for (int j = 0; j < numMetadata; j++) {
    node.metadata.add(data[i + j]);
  }
  
  return node;
}

int childSize(Node node) {
  int size = 2; // header size
  for (var child in node.children) {
    size += childSize(child);
  }
  size += node.metadata.length;
  return size;
}

void main() {
  List<int> data = File('input.txt').readAsStringSync().trim().split(' ').map(int.parse).toList();
  Node root = parseNode(data, 0);
  
  int sumMetadata = root.sumMetadata();
  int valueRoot = root.calculateValue();
  
  print(sumMetadata);
  print(valueRoot);
}

File: training_data/reflection-tuning/day8_part2_2018.clj
(ns day8
  (:require [clojure.string :as str]))

(defn parse-input [file]
  (->> (slurp file)
       (str/trim)
       (re-seq #"\d+")
       (mapv #(Integer/parseInt %))))

(defn process-node [numbers]
  (let [[child-count metadata-count & rest] numbers
        [children rest metadata sum-metadata] 
        (loop [i 0
               rest rest
               children []
               sum 0]
          (if (< i child-count)
            (let [[child rest' child-sum] (process-node rest)]
              (recur (inc i) rest' (conj children child) (+ sum child-sum)))
            [children 
             (drop metadata-count rest) 
             (take metadata-count rest)
             (+ sum (reduce + (take metadata-count rest)))]))
        node-value (if (zero? child-count)
                     (reduce + metadata)
                     (reduce + (for [idx metadata
                                     :when (<= 1 idx (count children))]
                                 (get-in children [(dec idx) :value] 0))))]
    [{:children children
      :metadata metadata
      :value node-value}
     rest
     (+ sum-metadata node-value)]))

(defn solve [input]
  (let [[root _ sum-metadata root-value] (process-node input)]
    (println "Part 1:" sum-metadata)
    (println "Part 2:" (:value root))))

(def input (parse-input "input.txt"))
(solve input)
